### PR TITLE
feat(ingest): bigquery/snowflake - Store last profile date in state

### PR DESCRIPTION
**Pull request by treff7es**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/state/redundant_run_skip_handler.py [hunk 1]
```diff
@@ -120,6 +120,17 @@ def update_state(
         cur_state.begin_timestamp_millis = start_time_millis
         cur_state.end_timestamp_millis = end_time_millis
 
+    def get_last_state(self) -> Optional[BaseUsageCheckpointState]:
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/state/redundant_run_skip_handler.py:123]: why was this needed?

**Review by hsheth2 — COMMENTED** [review]
overall the code is looking good, although adding a stateful ingestion handler still seems rather verbose

File: metadata-ingestion/src/datahub/ingestion/source/state/stateful_ingestion_base.py [hunk 2]
```diff
@@ -100,6 +102,59 @@ class StatefulIngestionConfigBase(
...
+    )
+
+    @root_validator(pre=False)
+    def stateful_option_validator(cls, values: Dict) -> Dict:
+        sti = values.get("stateful_ingestion")
+        if not sti or not sti.enabled:
+            if values.get("enable_lineage_extraction_state"):
+                logger.warning(
+                    "Stateful ingestion is disabled, disabling enable_lineage_extraction_state config option as well"
+                )
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/state/stateful_ingestion_base.py:118]: looks like we're not updating `values` here

File: metadata-ingestion/src/datahub/ingestion/source/state/profiling_state_handler.py [hunk 3]
```diff
@@ -21,12 +21,9 @@
 
 class StatefulProfilingConfig(StatefulIngestionConfig):
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/state/profiling_state_handler.py:22]: what do we use this class for?

Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/source/state/profiling_state_handler.py:22]: seems like we should be using `ProfilingStatefulIngestionConfig` instead?

**Review by hsheth2 — APPROVED** [review]

File: metadata-ingestion/src/datahub/ingestion/source/sql/sql_generic_profiler.py [hunk 4]
```diff
@@ -118,15 +125,34 @@ def is_dataset_eligible_for_profiling(
...
+        )
         if self.config.profiling.profile_if_updated_since_days is not None:
             threshold_time = datetime.now(timezone.utc) - timedelta(
                 self.config.profiling.profile_if_updated_since_days
             )
 
-        if not self.config.table_pattern.allowed(
-            dataset_name
-        ) or not self.config.profile_pattern.allowed(dataset_name):
+        if not self.config.profile_pattern.allowed(dataset_name):
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/sql/sql_generic_profiler.py:155]: Why is `self.config.table_pattern.allowed` check removed ?

File: metadata-ingestion/src/datahub/ingestion/source/sql/sql_generic_profiler.py [hunk 5]
```diff
@@ -118,15 +125,34 @@ def is_dataset_eligible_for_profiling(
...
+                # If profiling state exists we have to carry over to the new state
+                self.state_handler.add_to_state(dataset_urn, last_profiled)
+
+        threshold_time: Optional[datetime] = (
+            datetime.fromtimestamp(last_profiled / 1000, timezone.utc)
+            if last_profiled
+            else None
+        )
         if self.config.profiling.profile_if_updated_since_days is not None:
             threshold_time = datetime.now(timezone.utc) - timedelta(
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/sql/sql_generic_profiler.py:151]: I think, if threshold_time (last profiling time for a table) is available in profiling state, it would be smarter not to override it by threshold_time computed from `profile_if_updated_since_days`.

File: metadata-ingestion/src/datahub/ingestion/source/sql/sql_generic_profiler.py (hunk 4 above)
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/sql/sql_generic_profiler.py:155]: This was moved to the top to return early in this case

File: metadata-ingestion/src/datahub/ingestion/source/sql/sql_generic_profiler.py (hunk 5 above)
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/sql/sql_generic_profiler.py:151]: thanks, I will address this
